src/braket/experimental/autoqasm/api.py

```diff
@@ -320,9 +320,18 @@ def _convert_program_as_subroutine(
         root_oqpy_program._add_subroutine(subroutine_name, subroutine_function_call.subroutine_decl)
 
 
-def _make_return_instance(
-    f: Callable, program_conversion_context: aq_program.ProgramConversionContext
-) -> Any:
+def _make_return_instance_from_oqpy_return_type(return_type) -> Any:
+    if not return_type:
+        return None
+
+    return_type = aq_types.conversions.var_type_from_ast_type(return_type)
+    if return_type == aq_types.ArrayVar:
+        return []
+    return return_type()
+
+
+def _make_return_instance_from_f_annotation(f: Callable) -> Any:
+    # TODO: Recursive functions should work even if the user's type hint is wrong
     annotations = f.__annotations__
     return_type = annotations["return"] if "return" in annotations else None
```

Comment:
Need to add coverage for this func

Reply:
There is something weird about the code coverage detection for this codepath. We had to add `#pragma: no cover` inside `f_dummy` below, even though that code is actually hit during tests (e.g., if you put `raise NotImplementedError` inside `f_dummy`, it causes tests to fail). So I'd be curious to know if this code is actually not hit during tests, or if it's just a bug in the coverage detection.

Reply:
It used to get called for any subroutine, now it's only called for recursive subroutines. I had to add a new test with a recursive subroutine that returned a `List`